lib/SILOptimizer/Utils/ConstantFolding.cpp

```diff
@@ -1746,11 +1746,13 @@ ConstantFolder::processWorkList() {
                   if (OwnershipKind.isCompatibleWith(
                           ValueOwnershipKind::Guaranteed)) {
                     SILValue DTIResult = DTI->getResult(O->getOperandNumber());
-                    DTIResult->replaceAllUsesWith(NewVal);
-                    FoldedUsers.insert(DTI);
-                    if (auto *Inst = NewVal->getDefiningInstruction())
-                      WorkList.insert(Inst);
-                    continue;
+                    if (!DTIResult->use_empty()) {
```

Comment:
Reading the code, I would have no idea why you check for an empty use list. It needs a comment explaining that after the first RAUW of DTIResult, DTI will end up back on the worklist because it still uses the constant value. The second time around, that result should already have been RAUW'd so this breaks the infinite cycle.